scrape.js

```diff
@@ -20,47 +20,41 @@ const parseNameString = name => {
 
 const validateOpts = opts => {
     if (!Number.isInteger(opts.orgID)) {
-        return new Error("Invalid `orgID` field - not an integer")
+        throw new Error("Invalid `orgID` field - not an integer")
     }
 
     if (!Number.isInteger(opts.groupID)) {
-        return new Error("Invalid `groupID` field - not an integer")
+        throw new Error("Invalid `groupID` field - not an integer")
     }
 
     if ("debug" in opts) {
         const debug = opts.debug;
         if (debug !== undefined && debug !== false && debug !== true) {
-            return new Error("Invalid `debug` field - expected `undefined`, `false`, or `true`")
+            throw new Error("Invalid `debug` field - expected `undefined`, `false`, or `true`")
         }
     }
 
     if ("auth" in opts === false) {
-        return new Error("Missing `auth` object")
+        throw new Error("Missing `auth` object")
     } else if (typeof opts.auth.email !== "string") {
-        return new Error(`Invalid \`auth.email\` field - expected "string", got "${typeof opts.auth.email}"`)
+        throw new Error(`Invalid \`auth.email\` field - expected "string", got "${typeof opts.auth.email}"`)
     } else if (typeof opts.auth.password !== "string") {
-        return new Error(`Invalid \`auth.password\` field - expected "string", got "${typeof opts.auth.password}"`)
+        throw new Error(`Invalid \`auth.password\` field - expected "string", got "${typeof opts.auth.password}"`)
     }
-
-    return null;
 }
 
-module.exports = (opts = {}) => {
+module.exports = async (opts = {}) => {
+    validateOpts(opts)
```

Comment:
My reasoning is that `validateOpts` is purely for determining whether or not a give object is valid or not. In a similar fashion you wouldn't want `str.contains(substring)` throwing an error if `substring` wasn't inside the `str`.

Here, the main exported function decides that it should directly throw the error.

Also, having `validateOpts` not throw an error would be have more idiomatically when implementing https://github.com/compsoc-edinburgh/eusa-members-api/pull/1#discussion_r336715809 (if it would happen).

I think a better explanation would be `validateOpts` isn't failing in its job, it's just reporting a success or an error, and so it shouldn't throw anything.

> but another option is that `validateOpts` returns an error variable including the message(s) of the errors

If what you mean is just returning a `string`, that would work, but imo this feels nicer.

Reply:
I really disagree with returning errors - we’d just be catching them to rethought them, which would be more lines of code. The style of passing around error return values feels more like idiomatic Go or Rust, not JavaScript, and I think if we were going to use custom error handling behaviour, we would have to implement it everywhere. For me, something like validateOpts sounds like it should fail early by throwing errors. If error were wanted to be returned, it could maybe by called getOptsValidationErrors, or something. I do think that throwing errors inside validateOpts is more idiomatic JavaScript though

Reply:
Another argument against just returning a `string` is that it's useful to return more semantic errors like `TypeError`, `RangeError`, or other custom errors. That's not something the caller function can (or should) determine from the string alone.

Doing something like`getOptsValidationErrors` might work a little, but throwing multiple errors isn't possible in JavaScript, so might have little benefit here. There's ways to work around that, though.

----

When it comes down to it, I think that `validateOpts` _isn't_ failing in its job, and therefore shouldn't "crash" (by throwing an error).

The way I see `validateOpts` is that it's a helper function to determine whether or not options is _correct_, and not simply an extension of the logic from the main `scrape_members` export.

So I guess I disagree with changing the purpose of the function (from a helper vs. an 'extension of `scrape_members`').

I don't have strong feelings either way but I would still appreciate another point of view, so I've re-requested @pkage's review so they could make the deciding choice on this.